File: src/Model/ProgramState.java
package Model;

import Model.FileStatement.FileData;
import Model.FileStatement.IFileTable;
import Model.GeneralStatement.IStatement;
import Model.Utils.IDictionary;
import Model.Utils.IHeap;
import Model.Utils.IList;
import Model.Utils.IStack;


public class ProgramState {
    IStatement originalProgram;
    private IStack<IStatement> executionStack;
    private IDictionary<String, Integer> symTable;
    private IList<Integer> out;
    private IFileTable<Integer, FileData> fileTable;
    private IHeap<Integer, Integer> heap;

    public ProgramState(IStack<IStatement> executionStack, IDictionary<String, Integer> symTable, IList<Integer> out, IStatement originalProgram, IHeap<Integer, Integer> heap) {
        this.executionStack = executionStack;
        this.symTable = symTable;
        this.out = out;
        this.originalProgram = originalProgram;
        this.heap = heap;
    }

    public ProgramState(IStack<IStatement> executionStack, IDictionary<String, Integer> symTable, IList<Integer> out) {
        this.executionStack = executionStack;
        this.symTable = symTable;
        this.out = out;
    }

    public ProgramState(IStatement originalProgram, IStack<IStatement> executionStack, IDictionary<String, Integer> symTable, IList<Integer> out, IFileTable<Integer, FileData> fileTable, IHeap<Integer, Integer> heap) {
        this.originalProgram = originalProgram;
        this.executionStack = executionStack;
        this.symTable = symTable;
        this.out = out;
        this.fileTable = fileTable;
        this.heap = heap;
        executionStack.push(originalProgram);
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("ExecutionStack:");

        for (IStatement statement : executionStack.getAll()) {
            builder.append("\n\t").append(statement.toString());
        }
        builder.append("\nSymTable:");
        for (String key : symTable.getAll())
            builder.append("\n\t").append(key).append("::=").append(symTable.getValue(key));
        builder.append("\n").append(heap);
        builder.append("\nFileTable:");
        builder.append("\t").append(fileTable.toString());
        builder.append("\nOut:");
        for (Integer i : out.getAll())
            builder.append("\n\t").append(i);
        return builder.toString();
    }

    public IStack<IStatement> getExecutionStack() {
        return executionStack;
    }

    public void setExecutionStack(IStack<IStatement> executionStack) {
        this.executionStack = executionStack;
    }

    public IDictionary<String, Integer> getSymTable() {
        return symTable;
    }

    public void setSymTable(IDictionary<String, Integer> symTable) {
        this.symTable = symTable;
    }

    public IList<Integer> getOut() {
        return out;
    }

    public void setOut(IList<Integer> out) {
        this.out = out;
    }

    public IStatement getOriginalProgram() {
        return originalProgram;
    }

    public IFileTable<Integer, FileData> getFileTable() {
        return fileTable;
    }

    public void setFileTable(IFileTable<Integer, FileData> fileTable) {
        this.fileTable = fileTable;
    }

    public void setOriginalProgram(IStatement originalProgram) {
        this.originalProgram = originalProgram;
    }

    public IHeap<Integer, Integer> getHeap() {
        return heap;
    }

    public void setHeap(IHeap<Integer, Integer> heap) {
        this.heap = heap;
    }
}

File: Toy-Language-C#-Intepreter/Toy-Language-CS-Interpreter/Toy-Language-CS-Interpreter/Model/ProgramState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Toy_Language_CS_Interpreter.Model.Containers;
using Toy_Language_CS_Interpreter.Model.Statement;
using Toy_Language_CS_Interpreter.Model.Statement.FileStatement;

namespace Toy_Language_CS_Interpreter.Model
{
    class ProgramState
    {
        private Containers.IDictionary<string, int> symbolTable;
        private IStack<IStatement> executionStack;
        private Containers.IList<int> output;
        private IFileTable<int, FileData> fileTable;

        public ProgramState(Containers.IDictionary<string, int> symbolTable, IStack<IStatement> executionStack, Containers.IList<int> output, IFileTable<int, FileData> fileTable)
        {
            this.symbolTable = symbolTable;
            this.executionStack = executionStack;
            this.output = output;
            this.fileTable = fileTable;
        }


        public Containers.IDictionary<string, int> SymbolTable
        {
            get { return symbolTable; }
            set { symbolTable = value; }
        }

        public IStack<IStatement> ExecutionStack
        {
            get { return executionStack; }
            set { executionStack = value; }
        }

        public Containers.IList<int> Output
        {
            get { return output; }
            set { output = value; }
        }

        public IFileTable<int, FileData> FileTable
        {
            get { return fileTable; }
            set { fileTable = value; }
        }

        public override string ToString()
        {
            StringBuilder stringBuilder = new StringBuilder();
            stringBuilder.Append("ExecutionSack: \r\n");
            stringBuilder.Append(executionStack).Append("\r\n");
            stringBuilder.Append("Symbol Table:\r\n");
            stringBuilder.Append(symbolTable).Append("\r\n");
            stringBuilder.Append("FileTable:\r\n");
            stringBuilder.Append(fileTable).Append("\r\n");
            stringBuilder.Append("Output:\r\n");
            stringBuilder.Append(output).Append("\r\n");

            return stringBuilder.ToString();
        }

    }
}
